### PR TITLE
[hydrawise] Migrate to new GraphQL based API

### LINKED ISSUE #7261: [hydrawise] Stuck in Initializing phase
HI, I have openHAB 2.5 on Raspbian, my hydrawise binding added successfully, however when I add a Thing HydraWise Cloud it stucks in Initializing status. I have checked the API key and it works, nothing shows up in the logs

### PR DESCRIPTION
**Pull request by digitaldan**
Fixes #7261

This is a major refactoring of the cloud communication logic due to  Hydrawise migrating to a new GraphQL based API about a year ago, breaking the existing binding.   Users have been manually installing and testing this branch for about a year as i have dragged my feet getting a PR submitted. 

Signed-off-by: Dan Cunningham <dan@digitaldan.com>

### CODE REVIEW

File: bundles/org.openhab.binding.hydrawise/src/main/java/org/openhab/binding/hydrawise/internal/discovery/HydrawiseCloudControllerDiscoveryService.java [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+    }
+
+    private void addDiscoveryResults(Controller controller) {
+        String label = String.format("Hydrawise Controller %s", controller.name);
+        int id = controller.id;
+        ThingUID bridgeUID = bridge.getThing().getUID();
+        ThingUID thingUID = new ThingUID(HydrawiseBindingConstants.THING_TYPE_CONTROLLER, bridgeUID,
+                String.valueOf(id));
+
+        thingDiscovered(DiscoveryResultBuilder.create(thingUID).withLabel(label).withBridge(bridgeUID)
```
Comment (fwolter) [inline bundles/org.openhab.binding.hydrawise/src/main/java/org/openhab/binding/hydrawise/internal/discovery/HydrawiseCloudControllerDiscoveryService.java:73]: Can you add the representation property?

File: bundles/org.openhab.binding.hydrawise/src/main/java/org/openhab/binding/hydrawise/internal/handler/HydrawiseAccountHandler.java [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+                oAuthService.getAccessTokenByResourceOwnerPasswordCredentials(config.userName, config.password, SCOPE);
+            } else if (oAuthService.getAccessTokenResponse() == null) {
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, "Login credentials required.");
+                return;
+            }
+            this.refresh = Math.max(config.refreshInterval, MIN_REFRESH_SECONDS);
+            initPolling(0, refresh);
+        } catch (OAuthException | IOException e) {
+            logger.debug("Could not log in", e);
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, e.getMessage());
```
Comment (fwolter) [inline bundles/org.openhab.binding.hydrawise/src/main/java/org/openhab/binding/hydrawise/internal/handler/HydrawiseAccountHandler.java:144]: The log message could be removed, as the status update is already logged by the framework.

The state change originated by updateStatus() is logged to events.log. Including the status detail message.

File: bundles/org.openhab.binding.hydrawise/src/main/resources/OH-INF/thing/channel-types.xml [hunk 3]
```diff
@@ -49,13 +53,51 @@
...
+	<channel-type id="lastcontacttime" advanced="true">
+		<item-type>DateTime</item-type>
+		<label>Last Contact Time</label>
+		<description>Last contact time of a controller</description>
+		<state readOnly="true"></state>
+	</channel-type>
+
+	<channel-type id="summary">
+		<item-type>String</item-type>
+		<label>Status summary</label>
```
Comment (fwolter) [inline bundles/org.openhab.binding.hydrawise/src/main/resources/OH-INF/thing/channel-types.xml:87]:
```suggestion
		<label>Status Summary</label>
```

File: bundles/org.openhab.binding.hydrawise/src/main/resources/OH-INF/thing/channel-types.xml [hunk 4]
```diff
@@ -49,13 +53,51 @@
...
+		</channels>
+	</channel-group-type>
+
+	<channel-group-type id="system">
+		<label>System</label>
+		<description>Controller system data</description>
+		<channels>
+			<channel id="name" typeId="name"/>
+			<channel id="summary" typeId="summary"/>
+			<channel id="online" typeId="online"/>
```
Comment (fwolter) [inline bundles/org.openhab.binding.hydrawise/src/main/resources/OH-INF/thing/channel-types.xml:71]: Is this channel newly introduced or is this just moved? This seems rather the ThingStatus, than a channel.

File: bundles/org.openhab.binding.hydrawise/src/main/resources/OH-INF/thing/things.xml [hunk 5]
```diff
@@ -4,14 +4,50 @@
...
 
-	<!-- Sample Thing Type -->
-	<thing-type id="cloud">
-		<label>Hydrawise Cloud Thing</label>
-		<description>Hydrawise cloud connected irrigation system</description>
+	<bridge-type id="account">
+		<label>Hydrawise Account Thing</label>
+		<description>Hydrawise account</description>
+		<config-description>
+			<parameter name="userName" type="text" required="false">
```
Comment (fwolter) [inline bundles/org.openhab.binding.hydrawise/src/main/resources/OH-INF/thing/things.xml:11]: Isn't this required by intention?

File: bundles/org.openhab.binding.hydrawise/src/main/resources/OH-INF/thing/things.xml [hunk 6]
```diff
@@ -4,14 +4,50 @@
...
+				<context>password</context>
+				<description>Your Hydrawise account password, for security this will not be saved after the first login attempt
+					unless the "Save Password" option is enabled</description>
+			</parameter>
+			<parameter name="savePassword" type="boolean" required="false">
+				<label>Save Password</label>
+				<description>By default, the password will not be persisted after the first login attempt unless this is enabled</description>
+				<default>false</default>
+			</parameter>
+			<parameter name="refresh" type="integer" required="false" min="30">
```
Comment (fwolter) [inline bundles/org.openhab.binding.hydrawise/src/main/resources/OH-INF/thing/things.xml:26]:
```suggestion
			<parameter name="refresh" type="integer" required="false" min="30" unit="s">
```

**Review by fwolter — CHANGES_REQUESTED** [review]
There is one checkstyle warning left, which can be easily fixed.

File: bundles/org.openhab.binding.hydrawise/src/main/java/org/openhab/binding/hydrawise/internal/HydrawiseHandlerFactory.java [hunk 7]
```diff
@@ -70,4 +83,45 @@ public boolean supportsThingType(ThingTypeUID thingTypeUID) {
...
+    @Reference
+    protected void setOAuthFactory(OAuthFactory oAuthFactory) {
+        this.oAuthFactory = oAuthFactory;
+    }
+
+    protected void unsetOAuthFactory(OAuthFactory oAuthFactory) {
+        this.oAuthFactory = null;
+    }
+
+    private synchronized void registerDiscoveryService(HydrawiseAccountHandler handler) {
```
[the remark this answers is not in the document]
Comment (digitaldan) [reply bundles/org.openhab.binding.hydrawise/src/main/java/org/openhab/binding/hydrawise/internal/HydrawiseHandlerFactory.java]: old habits.... ThingHandlerService is much better, yes:-)

File: bundles/org.openhab.binding.hydrawise/src/main/resources/OH-INF/thing/channel-types.xml (hunk 4 above)
Comment (digitaldan) [reply bundles/org.openhab.binding.hydrawise/src/main/resources/OH-INF/thing/channel-types.xml]: Its new, and makes sense to adjust the ThingStatus, yes,  Thanks!

**Comment by digitaldan** [comment]
Thanks @fwolter as always for taking the time to review this PR!  I have addressed all your comments.

File: bundles/org.openhab.binding.hydrawise/src/main/java/org/openhab/binding/hydrawise/internal/handler/HydrawiseControllerHandler.java [hunk 8]
```diff
@@ -0,0 +1,430 @@
...
+
+        try {
+            switch (channelId) {
+                case CHANNEL_ZONE_RUN_CUSTOM:
+                    if (!(command instanceof QuantityType<?>)) {
+                        logger.warn("Invalid command type for run custom {}", command.getClass().getName());
+                        return;
+                    }
+                    if (allCommand) {
+                        client.runAllRelays(controllerId, ((QuantityType<?>) command).intValue());
```
Comment (fwolter) [inline bundles/org.openhab.binding.hydrawise/src/main/java/org/openhab/binding/hydrawise/internal/handler/HydrawiseControllerHandler.java:153]: You need to convert the value to the unit you expect with `toUnit()`.

File: bundles/org.openhab.binding.hydrawise/src/main/resources/OH-INF/thing/channel-types.xml [hunk 9]
```diff
@@ -98,13 +139,25 @@
...
+
+	<channel-type id="suspend">
+		<item-type>Switch</item-type>
+		<label>Suspend Zones</label>
+		<description>Suspends or resumes zones</description>
+	</channel-type>
+
+	<channel-type id="suspenduntil">
+		<item-type>DateTime</item-type>
+		<label>Suspend Zones </label>
```
Comment (fwolter) [inline bundles/org.openhab.binding.hydrawise/src/main/resources/OH-INF/thing/channel-types.xml:155]:
```suggestion
		<label>Suspend Zones</label>
```

File: bundles/org.openhab.binding.hydrawise/src/main/resources/OH-INF/thing/channel-types.xml [hunk 10]
```diff
@@ -220,4 +274,25 @@
...
+	<channel-type id="precipitation">
+		<item-type>Number</item-type>
+		<label>Precipitation</label>
+		<description>Precipitation amount</description>
+		<state readOnly="true" pattern="%.1f"/>
+	</channel-type>
+
+	<channel-type id="probabilityofprecipitation">
+		<item-type>Number</item-type>
+		<label>Probability Of Precipitation </label>
```
Comment (fwolter) [inline bundles/org.openhab.binding.hydrawise/src/main/resources/OH-INF/thing/channel-types.xml:294]:
```suggestion
		<label>Probability Of Precipitation</label>
```

**Review by fwolter — CHANGES_REQUESTED** [review]

File: bundles/org.openhab.binding.hydrawise/src/main/java/org/openhab/binding/hydrawise/internal/handler/HydrawiseControllerHandler.java (hunk 8 above)
Comment (digitaldan) [reply bundles/org.openhab.binding.hydrawise/src/main/java/org/openhab/binding/hydrawise/internal/handler/HydrawiseControllerHandler.java]: thanks for catching that.

**Comment by digitaldan** [comment]
Thanks @fwolter , i have addressed your latest review comments.

**Review by fwolter — APPROVED** [review]
LGTM

**Comment by digitaldan** [comment]
Thanks!

**Pull request merged**